plugins/org.python.pydev.core/pysrc/pydev_coverage.py

```diff
@@ -2,6 +2,25 @@
 Entry point module to run code-coverage.
 '''
 
+
+def is_valid_py_file(path):
+    '''
+    Checks whether the file can be read by the coverage module. This is especially
+    needed for .pyx files and .py files with syntax errors.
+    '''
+    import os
+
+    is_valid = False
+    if os.path.isfile(path) and not os.path.splitext(path)[1] == '.pyx':
+        try:
+            with open(path, 'r') as f:
+                compile(f.read(), path, 'exec')
```

Comment:
I was going to merge the pull request but something just occurred to me:

When you use `open(path, 'r')`, the contents will be converted to unicode using the current locale on python 3, which may not be what's expected (which is using the encoding declared on the file).

I.e.:
```
Python 3.6.5 |Anaconda, Inc.| (default, Mar 29 2018, 13:32:41) [MSC v.1900 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> x = open(r'c:\temp\foo.txt','r')
>>> x
<_io.TextIOWrapper name='c:\\temp\\foo.txt' mode='r' encoding='cp1252'>
```

I think `compile` accepts bytes and then does the right thing (uses the encoding declared on the file)... If it doesn't,  `tokenize.open` needs to be used instead of a plain `open`. 

Given that this is tricky and changes on py2/py3, it'd be nice to add such a test on a file with unicode chars which has some exoteric encoding -- i.e.: not utf-8 nor your locale encoding.

Also, it'd be nice if you could create a pull request in https://github.com/fabioz/PyDev.Debugger/ (python tests run only there and PyDev itself syncs back these files from that repo).

Reply:
Good point. I'm no so familiar with python 3. It's correct, compile with byte content does the trick.

About PyDev.Debugger:  What is the correct workflow? Right way would be to only create the pull request to PyDev.Debugger (which changes will be synced to this repo)? Is this pull request then kind of invalid?

Reply:
Yes, the proper way would be creating the pull request there (doing it here is not entirely wrong for simple changes though as I just make the sync the other way, but for anything more complex which requires tests doing it on PyDev.Debugger is the way to go as tests can be added and run on travis/appveyor).